src/main/java/de/tum/in/www1/artemis/service/ProgressBasedAchievementService.java

```diff
@@ -21,12 +25,58 @@
 
     private final static long EXERCISES_AMOUNT_BRONZE = 5L;
 
+    private final static long EXERCISES_AMOUNT_UNRANKED = 1L;
+
     private final static long MIN_SCORE_TO_QUALIFY = 50L;
 
-    public ProgressBasedAchievementService(AchievementRepository achievementRepository) {
+    public ProgressBasedAchievementService(StudentParticipationRepository studentParticipationRepository, AchievementRepository achievementRepository) {
+        this.studentParticipationRepository = studentParticipationRepository;
         this.achievementRepository = achievementRepository;
     }
 
+    /**
+     * Checks whether a user earned progress based achievements by counting all participations
+     * where the user scored at least as many points as defined in the achievement's minScoreToQualify property
+     * @param result the currently saved result
+     * @param course the given course
+     * @param user the given user
+     * @param achievements all progress based achievements within the given course
+     * @return the highest rank reached, returns null if no rank was reached
+     */
+    public AchievementRank checkForAchievement(Result result, Course course, User user, Set<Achievement> achievements) {
```

Comment:
```suggestion
    public Optional<AchievementRank> getAchievementRank(Result result, Course course, User user, Set<Achievement> achievements) {
```

By reading `check` in the method name, I would assume this method either returns a `boolean` or `throws` an exception if a check fails, so `get` might describe the purpose of this function a little better. 
Also, since there are multiple conditions making this method return `null`, using an `Optional` in the method signature clearly indicates the returned `AchievementRank` might not exist.

The same applies to the `checkForAchievement` methods in the other `[...]BasedAchievementService`s.

Reply:
That makes perfect sense, thanks for the advice :)
Refactored the functions accordingly.